lib/Sema/ConstraintGraph.h

```diff
@@ -202,20 +203,30 @@ class ConstraintGraph {
     return TypeVariables;
   }
 
+  /// Describes a single component, as produced by the connected components
+  /// algorithm.
+  struct Component {
+    /// The type variables in this component.
+    TinyPtrVector<TypeVariableType *> typeVars;
+
+    /// The constraints in this component.
+    TinyPtrVector<Constraint *> constraints;
+
+    /// Whether this component represents an orphaned constraint.
```

Comment:
It looks like you were considering returning the orphaned constraints as Components from connectedComponents() as well, since that would simplify the caller's logic even more, and then decided not to?

In any case, this method isn't used and Components won't be made with empty typeVars so its existence might cause confusion.

Reply:
You’re right! I considered that and decided that orphaned constraints were better handled by the caller. I’ll clean up this dead code once I get my performance-test results back